src/microbe_stage/MicrobeAI.cs

```diff
@@ -384,44 +411,94 @@ private void RunAndTumble(Random random)
+        foreach (var compoundWeight in compoundsSearchWeights)
         {
-            usefulCompounds = usefulCompounds.Where(x => x.Key != ammonia && x.Key != phosphates);
-        }
+            // Note this is about absorbed quantities (which is all microbe has access to) not the ones in the clouds.
+            // Gradient computation is therefore cell-centered, and might be different for different cells.
+            float compoundDifference = 0.0f;
 
-        float compoundDifference = 0.0f;
-        foreach (var compound in usefulCompounds)
-        {
-            compoundDifference += compound.Value;
+            float quantityAbsorbedThisStep, quantityAbsorbedPreviousStep;
+            microbe.TotalAbsorbedCompounds.TryGetValue(compoundWeight.Key, out quantityAbsorbedThisStep);
+            previouslyAbsorbedCompounds.TryGetValue(compoundWeight.Key, out quantityAbsorbedPreviousStep);
+
+            compoundDifference += quantityAbsorbedThisStep - quantityAbsorbedPreviousStep;
+
+            compoundDifference *= compoundWeight.Value;
+            gradientValue += compoundDifference;
         }
 
+        // Implement a detection threshold to possibly rule out too tiny variations
+        // TODO: possibly include cell capacity correction
+        float differenceDetectionThreshold = Constants.AI_GRADIENT_DETECTION_THRESHOLD;
+
         // If food density is going down, back up and see if there's some more
-        if (compoundDifference < 0 && random.Next(0, 10) < 9)
+        if (gradientValue < -differenceDetectionThreshold && random.Next(0, 10) < 9)
         {
             MoveWithRandomTurn(2.5f, 3.0f, random);
         }
 
         // If there isn't any food here, it's a good idea to keep moving
-        if (compoundDifference == 0 && random.Next(0, 10) < 5)
+        if (Math.Abs(gradientValue) <= differenceDetectionThreshold && random.Next(0, 10) < 5)
         {
             MoveWithRandomTurn(0.0f, 0.4f, random);
         }
 
-        // If positive last step you gained compounds, so let's stick around
-        if (compoundDifference > 0)
+        // If positive last step you gained compounds, so let's move toward the source
+        if (gradientValue > differenceDetectionThreshold)
         {
-            // There's a decent chance to turn most of the way around
+            // There's a decent chance to turn by 90° to explore gradient
+            // 180° is useless since previous position let you absorb less compounds already
             if (random.Next(0, 10) < 4)
             {
-                MoveWithRandomTurn(0.0f, 3.0f, random);
+                MoveWithRandomTurn(0.0f, 1.5f, random);
             }
         }
     }
 
+    /// <summary>
+    ///   Prioritizing compounds that are stored in lesser quantities.
+    ///   If ATP-producing compounds are low (less than half storage capacities),
+    ///   non ATP-related compounds are discarded.
+    ///   Updates compoundsSearchWeights class dictionary.
```

Comment:
It's not static, so it is not a class variable. Instead should be mentioned "instance field"